### PR TITLE
fix: fold takes initial value by closure rather than by value

### PR DESCRIPTION
**Pull request by zzlk**
fixes https://github.com/hydro-project/hydroflow/issues/903

### CODE REVIEW

File: hydroflow_lang/src/graph/ops/fold.rs [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
-    RANGE_0, RANGE_1,
+    DelayType, OpInstGenerics, OperatorCategory, OperatorConstraints, OperatorInstance,
+    OperatorWriteOutput, Persistence, WriteContextArgs, RANGE_0, RANGE_1,
 };
 use crate::diagnostic::{Diagnostic, Level};
 
 /// > 1 input stream, 1 output stream
 ///
-/// > Arguments: an initial value, and a closure which itself takes two arguments:
+/// > Arguments: two arguments, both closures. The first closure can be used to create the initial value for the accumulator, and the second itself takes two arguments:
```
Comment (zzlk) [inline hydroflow_lang/src/graph/ops/fold.rs:11]: this wording is bad

File: hydroflow_lang/src/graph/ops/fold.rs [hunk 2]
```diff
@@ -26,7 +25,7 @@ use crate::diagnostic::{Diagnostic, Level};
 /// ```hydroflow
 /// // should print `Reassembled vector [1,2,3,4,5]`
 /// source_iter([1,2,3,4,5])
-///     -> fold::<'tick>(Vec::new(), |accum: &mut Vec<_>, elem| {
+///     -> fold::<'tick>(|| Vec::new(), |accum: &mut Vec<_>, elem| {
```
Comment (zzlk) [inline hydroflow_lang/src/graph/ops/fold.rs:28]: change to passing function directly.

File: hydroflow_lang/src/graph/ops/fold.rs [hunk 3]
```diff
@@ -81,13 +80,12 @@ pub const FOLD: OperatorConstraints = OperatorConstraints {
...
         let (write_prologue, write_iterator, write_iterator_after) = match persistence {
-            // TODO(mingwei): Issues if initial value is not copy.
-            // TODO(mingwei): Might introduce the initial value multiple times on scheduling.
             Persistence::Tick => (
                 Default::default(),
                 quote_spanned! {op_span=>
                     let #ident = {
-                        let mut #accumulator_ident = #init;
+                        #[allow(clippy::redundant_closure_call)]
+                        let mut #accumulator_ident = (#init)();
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/fold.rs:88]: May be worth trying to store `#init` in its own name ident in the prologue, or put a TODO here stating to try that again

File: hydroflow_lang/src/graph/ops/fold.rs [hunk 4]
```diff
@@ -101,8 +99,9 @@ pub const FOLD: OperatorConstraints = OperatorConstraints {
             ),
             Persistence::Static => (
                 quote_spanned! {op_span=>
+                    #[allow(clippy::redundant_closure_call)]
                     let #folddata_ident = #hydroflow.add_state(
-                        ::std::cell::Cell::new(::std::option::Option::Some(#init))
+                        ::std::cell::Cell::new(::std::option::Option::Some((#init)()))
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/fold.rs:104]: same here

**Review by MingweiSamuel — APPROVED** [review]
Add a bang to the commit message `fix!: ...` and a `BREAKING CHANGE:` section to mark it as a breaking change

File: hydroflow_lang/src/graph/ops/fold.rs (hunk 3 above)
Comment (zzlk) [reply hydroflow_lang/src/graph/ops/fold.rs:88]: what would that give us?

**Pull request merged**